File: src/presentations/3DWeb/panel.ts
// Control-panel DOM: the stats readout, status pill, and button enable/disable states.
// Pure writes from the view's session state — no cube or view logic here.
import { Status } from './types'
import { CUBE_MOVE_BUTTONS, STATUS_META } from './config'

function timeText(elapsedMs: number): string {
  const ms = elapsedMs || 0
  const total = ms / 1000
  const mm = Math.floor(total / 60)
  const ss = Math.floor(total % 60)
  const t = Math.floor((ms % 1000) / 100)
  const ssStr = (mm > 0 && ss < 10 ? '0' : '') + ss
  return (mm > 0 ? mm + ':' : '') + ssStr + '.' + t
}

export function updateStats(elapsedMs: number, moveCount: number) {
  const time = document.getElementById('stat-time')
  const moves = document.getElementById('stat-moves')
  if (time) time.textContent = timeText(elapsedMs)
  if (moves) moves.textContent = String(moveCount)
}

export function updateStatus(status: Status) {
  const st = STATUS_META[status]
  const pill = document.getElementById('status-pill')
  const dot = document.getElementById('status-dot')
  const label = document.getElementById('status-label')
  if (dot) {
    dot.style.background = st.color
    dot.style.boxShadow = '0 0 10px ' + st.color
  }
  if (label) {
    label.textContent = st.label
    label.style.color = st.color
  }
  if (pill) pill.style.borderColor = st.color
}

// While the solver is active, every control is locked except Abort. Abort is live only during a
// solve and shows a spinner + "Aborting…" while the abort unwinds (which isn't instant).
export function updateControls(solving: boolean, aborting: boolean) {
  const setDisabled = (id: string, disabled: boolean) => {
    const el = document.getElementById(id) as HTMLButtonElement | null
    if (el) el.disabled = disabled
  }
  const lockable = ['btn-scramble', 'btn-reset', 'solve', 'btn-recenter'].concat(
    Object.keys(CUBE_MOVE_BUTTONS)
  )
  lockable.forEach((id) => setDisabled(id, solving))
  const abort = document.getElementById('abort-solve') as HTMLButtonElement | null
  if (abort) {
    abort.disabled = !solving || aborting
    abort.classList.toggle('btn--busy', aborting)
    abort.textContent = aborting ? 'Aborting…' : 'Abort'
  }
}

File: android/app/src/main/java/com/mikeb/simplepuzzlecube/ui/Panel.kt
// The control panel — analog of ../src/presentations/3DWeb/panel.ts + the button rows
// from 3DWeb.html: stats readout, status pill, drivers (Scramble/Solve/Abort/Reset),
// manual face turns, and whole-cube re-orientations. Pure renders of UiState; every
// intent is a ViewModel call.
package com.mikeb.simplepuzzlecube.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Button
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp

@Composable
fun StatusRow(state: UiState, modifier: Modifier = Modifier) {
    val meta = STATUS_META.getValue(state.status)
    val color = meta.color ?: MaterialTheme.colorScheme.onSurface
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(modifier = Modifier.size(10.dp).clip(CircleShape).background(color))
            Text(meta.label, color = color, modifier = Modifier.padding(start = 8.dp))
        }
        Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
            Text("⏱ ${timeText(state.elapsedMs)}")
            Text("↻ ${state.moveCount}")
        }
    }
}

@Composable
fun DriverControls(
    state: UiState,
    onScramble: () -> Unit,
    onSolve: () -> Unit,
    onAbort: () -> Unit,
    onReset: () -> Unit,
    modifier: Modifier = Modifier
) {
    // While the solver is active, every control is locked except Abort; Abort is live only
    // during a solve and shows "Aborting…" while the abort unwinds (which isn't instant).
    val locked = state.solverActive
    val slim = androidx.compose.foundation.layout.PaddingValues(horizontal = 4.dp, vertical = 8.dp)
    Row(modifier = modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Button(onClick = onScramble, enabled = !locked, modifier = Modifier.weight(1f), contentPadding = slim) {
            Text("Scramble", maxLines = 1)
        }
        if (state.solverActive) {
            Button(onClick = onAbort, enabled = !state.aborting, modifier = Modifier.weight(1f), contentPadding = slim) {
                Text(if (state.aborting) "Aborting…" else "Abort", maxLines = 1)
            }
        } else {
            Button(onClick = onSolve, modifier = Modifier.weight(1f), contentPadding = slim) {
                Text("Solve", maxLines = 1)
            }
        }
        OutlinedButton(onClick = onReset, enabled = !locked, modifier = Modifier.weight(1f), contentPadding = slim) {
            Text("Reset", maxLines = 1)
        }
    }
}

@Composable
fun ManualControls(
    state: UiState,
    prime: Boolean,
    onPrimeChange: (Boolean) -> Unit,
    onFaceMove: (MoveKey) -> Unit,
    onCubeMove: (CubeMoveKey) -> Unit,
    modifier: Modifier = Modifier
) {
    val locked = state.solverActive
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            listOf(MoveKey.U, MoveKey.D, MoveKey.L, MoveKey.R, MoveKey.F, MoveKey.B).forEach { key ->
                OutlinedButton(
                    onClick = { onFaceMove(key) },
                    enabled = !locked,
                    modifier = Modifier.weight(1f),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(0.dp)
                ) {
                    Text(key.name + if (prime) "′" else "")
                }
            }
            FilterChip(selected = prime, onClick = { onPrimeChange(!prime) }, label = { Text("′") })
        }
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            listOf(
                "⟲ Spin" to CubeMoveKey.spinLeft,
                "⟳ Spin" to CubeMoveKey.spinRight,
                "↑ Roll" to CubeMoveKey.rollUp,
                "↓ Roll" to CubeMoveKey.rollDown
            ).forEach { (label, key) ->
                OutlinedButton(
                    onClick = { onCubeMove(key) },
                    enabled = !locked,
                    modifier = Modifier.weight(1f),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(0.dp)
                ) {
                    Text(label)
                }
            }
        }
    }
}
